feat: modify L2 output submitter module corresponding to the changed contracts

seolaoh opened the pull request:
Check if selected as next validator when submitting L2 output and only submit if selected,
and add bond amount to output submission transaction data.
Also to reduce unnecessary requests, change from repeating the submission attempt every poll interval
to waiting L2 blocks until the next submission before retrying submission.

Add `AccessList` to `TxCandidate` to reduce gas price if set. 
Set `AccessList` to `ValidatorPool` contract address and storage key of it to access when submitting L2 output transaction.

Related: https://github.com/kroma-network/kroma/issues/19

seolaoh commented on ops-devnet/devnet-up.sh:
```diff
@@ -97,14 +96,16 @@ fi
 
 L2OO_ADDRESS="0x6900000000000000000000000000000000000004"
 COLOSSEUM_ADDRESS="0x690000000000000000000000000000000000000D"
+VP_ADDRESS=""
```
We need to set this address, but I'm not sure what address to set it so I left it empty.

seolaoh commented on ops-devnet/docker-compose.yml:
```diff
@@ -115,9 +115,11 @@ services:
       VALIDATOR_PPROF_ENABLED: "true"
       VALIDATOR_METRICS_ENABLED: "true"
       VALIDATOR_ALLOW_NON_FINALIZED: "true"
+      VALIDATOR_OUTPUT_SUBMITTER_BOND_AMOUNT: 1
```
This value needs to be determined, I arbitrarily set it to 1.

seolaoh commented on components/validator/validator.go:
```diff
@@ -129,23 +130,23 @@ func (v *Validator) Stop() {
+				return
 			}
-
-			if err := v.submitChallengeTx(); err != nil {
-				v.l.Error("failed to submit challenge tx", "err", err)
-			}
```
Submitting challenge is about to implemented in another PR outside of this `for` statement, so I remove it.

seolaoh commented:
I added some changes described below.
- aligned the structure of the `l2_output_submitter.go` with the `challenger.go` structure implemented in https://github.com/kroma-network/kroma/pull/59 in a uniform way.
- changed the way to wait for the next output submission from using `time.Sleep` to using `time.AfterFunc`.
- removed the storage key representing the slot of `ValidatorPool.bonds` length. To add proper storage key, we need to calculate exact position of `ValidatorPool.bonds[outputIndex]`, but we do not fetch `outputIndex` currently so I didn't add it.

chokobole commented on components/validator/l2_output_submitter.go:
```diff
@@ -37,33 +57,130 @@ func NewL2OutputSubmitter(cfg Config, l log.Logger) (*L2OutputSubmitter, error)
+
+func (l *L2OutputSubmitter) fetchBlockNumbers(ctx context.Context) (*big.Int, *big.Int, error) {
+	cCtx, cancel := context.WithTimeout(ctx, l.cfg.NetworkTimeout)
+	defer cancel()
```
I think it's better to explicitly call `cancel()` for each position.

chokobole commented on components/validator/validator.go:
```diff
@@ -86,7 +87,9 @@ func NewValidator(parentCtx context.Context, cfg Config, l log.Logger, m metrics
 	ctx, cancel := context.WithCancel(parentCtx)
 
-	l2OutputSubmitter, err := NewL2OutputSubmitter(cfg, l)
+	txCandidatesChan := make(chan txmgr.TxCandidate, 10)
```
Is there any possibility that it sends through the channel more than 10? How about injecting the value from the config or something?

seolaoh replied:
Output submission transaction is sent hourly (if selected next validator always), and challenge is a rare case, so I think it would be ok now. wdyt?

chokobole replied:
I was just worried that there's case this number is not enough for e2e or something. Let's tackle the issue when we face it!

chokobole reviewed (approved):
LGTM

chokobole reviewed (approved):
LGTM

chokobole reviewed (approved):
LGTM

0xHansLee commented on components/validator/l2_output_submitter.go:
```diff
@@ -37,33 +57,131 @@ func NewL2OutputSubmitter(cfg Config, l log.Logger) (*L2OutputSubmitter, error)
+
+func (l *L2OutputSubmitter) fetchBlockNumbers(ctx context.Context) (*big.Int, *big.Int, error) {
+	cCtx, cancel := context.WithTimeout(ctx, l.cfg.NetworkTimeout)
+	callOpts := utils.NewCallOptsWithSender(cCtx, l.cfg.TxManager.From())
```
Instead of declaring `callOpts` in each function, why don't we reuse it with declaring in `L2OutputSubmitter`, as done in [`Challenger`](https://github.com/kroma-network/kroma/blob/9b2e8984db0b74dd267aa87119bab4b92354cc6c/components/validator/challenger.go#L52)?

seolaoh replied:
As explained offline, it's because we're using `callOpts` with network timeout context. We need to declare the context for each network call.

seolaoh commented on components/validator/flags/flags.go:
```diff
@@ -68,6 +74,12 @@ var (
+		Usage:  "Amount to bond when submitting each output",
+		EnvVar: kservice.PrefixEnvVar(envVarPrefix, "OUTPUT_SUBMITTER_BOND_AMOUNT"),
+		Value:  1,
+	}
```
I added `(in wei)`.

seolaoh commented on components/validator/l2_output_submitter.go:
```diff
@@ -73,36 +191,70 @@ func (l *L2OutputSubmitter) FetchNextOutputInfo(ctx context.Context) (*eth.Outpu
+	l2BlockTime, err := l.l2ooContract.L2BLOCKTIME(callOpts)
+	if err != nil {
+		l.log.Error("validator unable to get L2 block time", "err", err)
+		return err
```
I added to fetch `l2BlockTime` and `submissionInterval` from contract call and add it `L2OutputSubmitter` in func `NewL2OutputSubmitter`. We need to move this part to validator config maybe later to use it in challenger too.

seolaoh commented:
I added some new changes described below.
- in func `NewL2OutputSubmitter`, cancel the created context when error occurred
- added `retryAfter` func
- in func `loop`, trigger `submitChan` after 1 second when error occurred in executing `trySubmitL2Output`
- in func `trySubmitL2Output`, trigger `submitChan` after 1 second when succeed to submit output
- in func `canSubmit`, when not selected next validator, wait 1800 more blocks
- in func `waitL2Blocks`, if `currentBlockNumber` is bigger than `targetBlockNumber`, wait just 1 second and trigger `submitChan`
- in func `fetchOutput`, removed code block to compare current output block number with finalized or safe block number, because it's redundant part already done in `canSubmit`

do we need to set above 1 second using config?

0xHansLee commented on components/validator/l2_output_submitter.go:
```diff
@@ -3,67 +3,222 @@ package validator
+		return nil, err
+	}
+
+	cCtx, cCancel := context.WithTimeout(ctx, cfg.NetworkTimeout)
```
With this changes, please delete `cCancel()` in L78
```suggestion
	cCtx, cCancel := context.WithTimeout(ctx, cfg.NetworkTimeout)
	defer cCancel()
```

0xHansLee commented on components/validator/l2_output_submitter.go:
```diff
@@ -3,67 +3,222 @@ package validator
+	}
+	// Wait for L2 blocks proceeding until next submission is available when not selected for next validator
+	if !isNextValidator {
+		nextBlockNumberToWait = nextBlockNumberToWait.Add(nextBlockNumberToWait, l.submissionInterval)
```
```suggestion
		nextBlockNumberToWait = new(big.Int).Add(nextBlockNumberToWait, l.submissionInterval)
```

0xHansLee commented on components/validator/l2_output_submitter.go:
```diff
@@ -3,67 +3,222 @@ package validator
+}
+
+func (l *L2OutputSubmitter) fetchBlockNumbers(ctx context.Context) (*big.Int, *big.Int, error) {
+	cCtx, cancel := context.WithTimeout(ctx, l.cfg.NetworkTimeout)
```
```suggestion
	cCtx, cCancel := context.WithTimeout(ctx, l.cfg.NetworkTimeout)
	defer cCancel()
```

0xHansLee commented on components/validator/l2_output_submitter.go:
```diff
@@ -3,67 +3,222 @@ package validator
-	syncStatusCtx, cancelSyncStatusCtx := context.WithTimeout(ctx, l.cfg.NetworkTimeout)
-	defer cancelSyncStatusCtx()
-	status, err := l.cfg.RollupClient.SyncStatus(syncStatusCtx)
+	cCtx, cancel = context.WithTimeout(ctx, l.cfg.NetworkTimeout)
```
```suggestion
	cCtx, cCancel = context.WithTimeout(ctx, l.cfg.NetworkTimeout)
```

0xHansLee commented on components/validator/l2_output_submitter.go:
```diff
@@ -73,60 +228,88 @@ func (l *L2OutputSubmitter) FetchNextOutputInfo(ctx context.Context) (*eth.Outpu
+}
+
+func (l *L2OutputSubmitter) checkNextValidator(ctx context.Context) (bool, error) {
+	cCtx, cancel := context.WithTimeout(ctx, l.cfg.NetworkTimeout)
```
```suggestion
	cCtx, cCancel := context.WithTimeout(ctx, l.cfg.NetworkTimeout)
```

0xHansLee commented on components/validator/l2_output_submitter.go:
```diff
@@ -73,60 +228,88 @@ func (l *L2OutputSubmitter) FetchNextOutputInfo(ctx context.Context) (*eth.Outpu
+// fetchOutput gets the output information to the corresponding block number.
+// It returns the output info if the output can be made, otherwise error.
+func (l *L2OutputSubmitter) fetchOutput(ctx context.Context, blockNumber *big.Int) (*eth.OutputResponse, error) {
+	cCtx, cancel := context.WithTimeout(ctx, l.cfg.NetworkTimeout)
```
```suggestion
	cCtx, cCancel := context.WithTimeout(ctx, l.cfg.NetworkTimeout)
```

0xHansLee reviewed (approved):
LGTM.
The remainings are not mandatory to change.

seolaoh commented:
@0xHansLee I applied the comments that I react with emoji.
